feat: Add support for parsing CSV files with missing values

diegokolena opened the pull request:
This commit adds support for parsing CSV files with missing values. Four new CSV files were added to the test data directory, each containing different types of missing values. The  file contains missing values in the second column, the  file contains missing values in both columns, the  file contains missing values in the second column and JSON strings in the third and fourth columns, and the  file contains missing values in the first column.

Additionally, the  file was modified to include a new function  that attempts to parse empty strings as , numbers, and JSON strings into Python objects. This function is used to normalize the missing values in the CSV files.

This feature will improve the robustness of the CSV parsing functionality and allow for better handling of missing values in dataframes.

### Linked issue(s)

### What change does this PR introduce and why?

### Please check if the PR fulfills these requirements

- [ ] Include reference to internal ticket and/or GitHub issue "Fixes #NNNN" (if applicable)
- [ ] Relevant tests for the changes have been added
- [ ] Relevant docs have been added / updated

diegokolena committed:
feat: Add support for parsing CSV files with missing values

This commit adds support for parsing CSV files with missing values. Four new CSV files were added to the test data directory, each containing different types of missing values. The  file contains missing values in the second column, the  file contains missing values in both columns, the  file contains missing values in the second column and JSON strings in the third and fourth columns, and the  file contains missing values in the first column.

Additionally, the  file was modified to include a new function  that attempts to parse empty strings as , numbers, and JSON strings into Python objects. This function is used to normalize the missing values in the CSV files.

This feature will improve the robustness of the CSV parsing functionality and allow for better handling of missing values in dataframes.

yifanwu-kolena commented on kolena/_utils/dataframes/transformers.py:
```diff
@@ -51,19 +54,97 @@ def _normalize_dicts(dict_list: List[Dict[str, Any]], sep: str = ".", max_level:
+    # Normalize all dictionaries to have all keys,
+    # filling missing ones with None and parsing existing ones
     for d in flattened_dicts:
         for key in all_keys:
-            if key not in d:
-                d[key] = None
+            d[key] = _try_parse(d.get(key, None))
```
what's the rational to add parse at here? `_normalize_dicts` is only used by `json_normalize`.

diegokolena committed:
Normalize dictionaries to have all keys, filling missing ones with None

yifanwu-kolena commented on kolena/_utils/dataframes/transformers.py:
```diff
@@ -60,10 +63,88 @@ def _normalize_dicts(dict_list: List[Dict[str, Any]], sep: str = ".", max_level:
     return flattened_dicts
 
 
+def _try_parse(value: Any) -> Any:
```
i think we need to use this function in `_deserialize_dataobject_str` or somewhere else

consoull replied:
Ya, this function should be used in https://github.com/kolenaIO/kolena/blob/trunk/kolena/io.py#L61

yifanwu-kolena commented on kolena/_utils/dataframes/transformers.py:
```diff
@@ -60,10 +63,88 @@ def _normalize_dicts(dict_list: List[Dict[str, Any]], sep: str = ".", max_level:
+
+def parse_cols(df: pd.DataFrame) -> pd.DataFrame:
+    """
+    Normalize strings in DataFrame columns into respective Python objects.
+    """
+    for col in df.columns:
+        if df[col].dtype == "object":
+            df[col] = df[col].apply(lambda x: _try_parse(x), convert_dtype=False)
+    return df
+
+
+def drop_unnamed(df: pd.DataFrame) -> pd.DataFrame:
+    """
+    Drops any columns in the DataFrame that have an 'Unnamed' prefix in their column name.
+    """
+    cols_to_drop = [col for col in df.columns if col.startswith("Unnamed")]
+    df.drop(columns=cols_to_drop, inplace=True)
+    return df
+
+
+def drop_invalid_ground_truth(df: pd.DataFrame) -> pd.DataFrame:
+    """
+    Drops the invalid 'ground_truth' column from the DataFrame.
+    """
+    GROUND_TRUTH_FIELD = "ground_truth"
+    if GROUND_TRUTH_FIELD in df.columns:
+        if not isinstance(df[GROUND_TRUTH_FIELD].iloc[0], dict):
+            df.drop(columns=[GROUND_TRUTH_FIELD], inplace=True)
+    return df
+
+
+def drop_invalid_metadata(df: pd.DataFrame) -> pd.DataFrame:
+    """
+    Drops the invalid 'metadata' column from the DataFrame.
+    """
+    METADATA_FIELD = "metadata"
+    if METADATA_FIELD in df.columns:
+        if not isinstance(df[METADATA_FIELD].iloc[0], dict):
+            df.drop(columns=[METADATA_FIELD], inplace=True)
+    return df
```
do we need these functions to be ported over for now?

diegokolena committed:
Using _try_parse on _deserialize_dataobject_str
Remove unecesary fn

diegokolena committed:
feat: Remove unnecessary CSV files

This commit removes several CSV files from the test data directory that are no longer needed. The files nan.csv, heterogeneous.csv, parse_json.csv, __init__.py, and unnamed.csv have been deleted.

These files were either redundant, outdated, or no longer relevant to the project. Removing them helps to declutter the repository and improve maintainability.

diegokolena committed:
feat: Improve parsing of CSV files with missing values

This commit adds support for parsing CSV files with missing values. It includes modifications to the `_try_parse` function to handle empty strings, numbers, and JSON strings. Additionally, four new CSV files were added to the test data directory, each containing different types of missing values.

These changes enhance the robustness of the CSV parsing functionality and improve the handling of missing values in dataframes.

consoull commented:
Can we update the IO tests to catch regressions here

diegokolena committed:
feat: Refactor _try_parse function to handle empty strings and improve CSV parsing

The _try_parse function in transformers.py has been modified to handle empty strings as None values. This change improves the parsing of CSV files with missing values. Additionally, support has been added for parsing numeric strings, JSON strings, and NumPy arrays. The test_transformers.py file includes a new test case for the _try_parse function.

diegokolena committed:
Merge branch 'trunk' into diego/kol-5863-added_new_fn
